diff --git a/evaluator.go b/evaluator.go
index ab9f687..47aedf0 100644
--- a/evaluator.go
+++ b/evaluator.go
@@ -5,7 +5,6 @@ import (
 	"fmt"
 	"io"
 	"log"
-	"os"
 
 	"github.com/charmbracelet/vhs/lexer"
 	"github.com/charmbracelet/vhs/parser"
@@ -116,11 +115,6 @@ func Evaluate(ctx context.Context, tape string, out io.Writer, opts ...Evaluator
 
 	// Clean up temporary files at the end.
 	defer func() {
-		if v.Options.Video.Output.Frames != "" {
-			// Move the frames to the output directory.
-			_ = os.Rename(v.Options.Video.Input, v.Options.Video.Output.Frames)
-		}
-
 		_ = v.Cleanup()
 	}()
 
diff --git a/vhs.go b/vhs.go
index 383b0cf..dcdd0f8 100644
--- a/vhs.go
+++ b/vhs.go
@@ -232,6 +232,7 @@ func (vhs *VHS) Render() error {
 	cmds = append(cmds, MakeMP4(vhs.Options.Video))
 	cmds = append(cmds, MakeWebM(vhs.Options.Video))
 	cmds = append(cmds, MakeScreenshots(vhs.Options.Screenshot)...)
+	cmds = append(cmds, MakeFrames(vhs.Options.Video, vhs.totalFrames)...)
 
 	for _, cmd := range cmds {
 		if cmd == nil {
@@ -239,6 +240,7 @@ func (vhs *VHS) Render() error {
 		out, err := cmd.CombinedOutput()
 		if err != nil {
 			log.Println(string(out))
+			return fmt.Errorf("render output: %w", err)
 		}
 	}
 
diff --git a/video.go b/video.go
index 8ed4929..36f54a1 100644
--- a/video.go
+++ b/video.go
@@ -166,3 +166,42 @@ func MakeWebM(opts VideoOptions) *exec.Cmd {
 func MakeMP4(opts VideoOptions) *exec.Cmd {
 	return makeMedia(opts, opts.Output.MP4)
 }
+
+func MakeFrames(opts VideoOptions, totalFrames int) []*exec.Cmd {
+	if opts.Output.Frames == "" {
+		return nil
+	}
+
+	ensureDir(filepath.Join(opts.Output.Frames, "frame.png"))
+	log.Println(GrayStyle.Render("Creating frames in " + opts.Output.Frames + "..."))
+	cmds := make([]*exec.Cmd, 0, totalFrames)
+	for frame := opts.StartingFrame; frame <= totalFrames; frame++ {
+		cursorStream := filepath.Join(opts.Input, fmt.Sprintf(cursorFrameFormat, frame))
+		textStream := filepath.Join(opts.Input, fmt.Sprintf(textFrameFormat, frame))
+		outputPath := filepath.Join(opts.Output.Frames, fmt.Sprintf("frame-%05d.png", frame))
+		cmds = append(cmds, exec.Command("ffmpeg", buildFrameFFopts(opts, outputPath, textStream, cursorStream)...))
+	}
+	return cmds
+}
+
+func buildFrameFFopts(opts VideoOptions, targetFile, textStream, cursorStream string) []string {
+	streamBuilder := NewStreamBuilder(2, opts.Input, opts.Style)
+	streamBuilder.args = append(streamBuilder.args,
+		"-y",
+		"-i", textStream,
+		"-i", cursorStream,
+	)
+	streamBuilder = streamBuilder.
+		WithMargin().
+		WithBar().
+		WithCorner()
+
+	filterBuilder := NewScreenshotFilterComplexBuilder(opts.Style).
+		WithWindowBar(streamBuilder.barStream).
+		WithBorderRadius(streamBuilder.cornerStream).
+		WithMarginFill(streamBuilder.marginStream)
+
+	args := streamBuilder.Build()
+	args = append(args, filterBuilder.Build()...)
+	return append(args, targetFile)
+}
